lib/Alchemy/Phrasea/PhraseanetService/Controller/PSExposeController.php

```diff
@@ -1284,48 +1313,132 @@ private function getAndSaveToken($exposeName)
 
-                if ($response->getStatusCode() !== 200) {
+                if ($resToken->getStatusCode() !== 200) {
                     return null;
                 }
 
-                $tokenBody = $response->getBody()->getContents();
+                $refreshtokenBody = $resToken->getBody()->getContents();
+
+                $refreshtokenBody = json_decode($refreshtokenBody,true);
+
+                if (isset($refreshtokenBody['refresh_expires_in'])) {
+                    $passSessionNameValue = [
+                        'access_token' => $refreshtokenBody['access_token'],
+                        'expires_at'   => time() + $refreshtokenBody['expires_in'],
+                        'refresh_token'=> $refreshtokenBody['refresh_token'],
+                        'refresh_expires_at' => time() + $refreshtokenBody['refresh_expires_in']
+                    ];
+                } else {
+                    $passSessionNameValue = [
+                        'access_token' => $refreshtokenBody['access_token'],
+                    ];
+                }
 
-                $tokenBody = json_decode($tokenBody,true);
+                $session->set($passSessionName, $passSessionNameValue);
 
-                $session->set($credentialSessionName, $tokenBody['access_token']);
+                $accessToken = $refreshtokenBody['access_token'];
+            } else {
+                $session->remove($passSessionName);
+            }
 
-                $accessToken = $tokenBody['access_token'];
+        } elseif ($config['connection_kind'] == 'client_credentials') {
+            if ($session->has($credentialSessionName)) {
+                $tokenInfoCredential = $session->get($credentialSessionName);
+                if (!isset($tokenInfoCredential['expires_at'])) {
+                    $accessToken = $tokenInfoCredential['access_token'];
+                } elseif (is_array($tokenInfoCredential) && $tokenInfoCredential['expires_at'] > time()) {
+                    $accessToken = $tokenInfoCredential['access_token'];
+                } else {
+                    $accessToken = $this->getTokenByCredential($oauthClient, $config, $credentialSessionName);
+                }
+            } else {
+                $accessToken = $this->getTokenByCredential($oauthClient, $config, $credentialSessionName);
             }
         }
 
         return $accessToken;
     }
 
+    private function getTokenByCredential(Client $oauthClient, array $exposeConfiguration, $credentialSessionName)
+    {
+        $session = $this->getSession();
+
+        $response = $oauthClient->post($exposeConfiguration['oauth_token_uri'] , [
+            'form_params' => [
+                'client_id'     => $exposeConfiguration['expose_client_id'],
+                'client_secret' => $exposeConfiguration['expose_client_secret'],
+                'grant_type'    => 'client_credentials',
+//                'scope'         => 'publish'
```

Comment:
Does it work if you uncomment?

Reply:
yes for v2 and v3

Reply:
OK so uncomment it :) so this client will explicitly request this scope.